Code migration guide for lakeFS v1 SDK (#6826)
treeverse/lakeFS: docs/project/code-migrate-1.0-sdk.md

---

> title: Code Migration Guide

Reviewer: ```suggestion
title: Migrating to 1.0
```

Reviewer 2: @ozkatz can we say "API" and/or "SDK" in the title somewhere? I want to make it clear to non-programmers that they don't need to read this. It will be more confusing to someone who just uses Spark/lakeFS.
```suggestion
title: Migrating to SDK version 1.0
```

Author: think we should keep it short as it will be on the left side menu

parent: The lakeFS Project
description: Code migration guide detailing API and SDK upgrades, deprecated and new API operations, and SDK migration processes for both Java/JVM and Python with code refactoring examples for improved stability and compatibility.

---

> # lakeFS v1 - Code Migration Guide

Reviewer: ```suggestion
# lakeFS 1.0 - Code Migration Guide
```

- `internalDeleteGarbageCollectionRules` (Deprecated. Temporary backward support. Alternative: `deleteGCRules`)
- `internalSetGarbageCollectionRules` (Deprecated. Temporary backward support. Alternative: `setGCRules`)
- `internalGetGarbageCollectionRules` (Deprecated. Temporary backward support. Alternative: `getGCRules`)
- `prepareGarbageCollectionCommits`
- `getGarbageCollectionConfig`

### New/Updated API Operations

Here are the newly added or updated operations:  

- `getConfig` (Retrieve lakeFS version and storage info)
- `setBranchProtectionRules` (Route updated)
- `getBranchProtectionRules` (Route updated)
- `getGCRules` (New route introduced)
- `setGCRules` (New route introduced)
- `deleteGCRules` (New route introduced)
- `importStatus` (Response structure updated: 'ImportStatusResp' to 'ImportStatus')
- `uploadObject` (Parameters 'if-none-match' and 'storageClass' are now deprecated)
- `prepareGarbageCollectionCommits` (Request body removed)
- `getOtfDiffs` & `otfDiff` (Removed from 'otf diff' tag; retained in 'experimental' tag)

## Migrating SDK Code for Java and JVM-based Languages

### Introduction

> If you are using the SDK for Java or any other JVM-based language, be aware that the current generated SDK is not stable with respect to minor version upgrades. Transitioning from `io.lakefs:lakefs-client` to `io.lakefs:sdk` will necessitate rewriting your API calls to fit the new design paradigm.

Reviewer 2: Can we use "lakeFS client" for the old version and "SDK" for the new?

Author: rewrote, use the term client and package

- **Improved Readability:** The fluent style makes it evident which parameters are required and which ones are optional, enhancing code clarity.

When migrating your code, ensure you refactor all your API calls to adopt the new fluent style. This ensures that your application remains maintainable and is safeguarded against potential issues arising from minor SDK version upgrades.

For an illustrative example of the transition between styles, you can view the changes made in this pull request: [lakeFS pull request #6529](https://github.com/treeverse/lakeFS/pull/6529/files#diff-4c50b9ac3bf6bfc05e3b6ff0fbe2fd3214f31afb5b449732d90efe5f97f67167R666).

## Migrating SDK Code for Python

### Introduction

If you are using the SDK for Python, be aware that the current generated SDK is not stable with respect to minor version upgrades. Transitioning from `lakefs-client` to `lakefs-sdk` will necessitate rewriting your API calls.

### Here's a breakdown of the changes:

1. **Modules change**
   - The previous `model` module was renamed to `models`, meaning that `lakefs_client.model` imports should be replaced with `lakefs_sdk.models` imports.

>    - In case the `LakeFSClient` class is not used (not recommended) and the `apis` module is used directly instead- the previous `apis` module doesn't exist. Instead, use the `api` module, meaning that `lakefs_client.apis` imports should be replaced with `lakefs_sdk.api` imports.

Reviewer 2: ```suggestion
   - If you chose to use the `apis` module is used directly without going through the `LakeFSClient` class: the previous `apis` module no longer exists. Instead, use the `api` module, meaning that `lakefs_client.apis` imports should be replaced with `lakefs_sdk.api` imports. We continue to recommend using lakefs_sdk.LakeFSClient.
```
